### PR TITLE
FlightGear simulator support

### PR DESCRIPTION
**Pull request by kaklik**
The PX4 stack seems does not currently support the [FlightGear](https://www.flightgear.org/) simulator for SITL simulation. The FlightGear has better support for modeling of rotorcraft than the current PX4's mainstream simulator Gazebo. 

This pull-request adding the possibility of the use of FlightGear(FG) and was tested on the [autogyro TF-G1 simulation model](https://github.com/ThunderFly-aerospace/FlightGear-TF-G1).  

It is a stand-alone application, which connect to FG (over UDP generic protocol) and PX4 ( over TCP, Mavlink)

image

image2

[The flight video](https://youtu.be/zcoMrHmb6sQ)

The flightgear_bridge has few limitations, which are not fully resolved: 

-  The PX4 is connected to FlightGear thought "[generic protocol](http://wiki.flightgear.org/Generic_protocol)", which is served synchronously to the simulator graphics engine frame rate. So the PX4 gets the sensor data in frequency, depending on graphics resources and the current scene.
 In the future, the bridge can upsample the data by generating noised samples from the last known value it is a terrible way. Question is: What is the lowest acceptable rate of sensors data? We did not find that.
- What do you think about the implementation of the auto-download of FG models for FlightGear? The suggestion is to use the submodules architecture to link PX4 simulation supported FG models. 
- The simulation must be run in no lockstep mode. 

To the simulation could be run by 

`make px4_sitl_nolockstep flightgear_rascal` for plane  or `make px4_sitl_nolockstep flightgear_tf-g1` for autogyro.

### CODE REVIEW

**Comment by TSC21** [comment]
This is really cool! Although, i do think it makes mores sense that thr flightgear_bridge is a repository by itself and that it gets included as a submodule here, similarly to sitl_gazebo and jMavSim.

**Comment by dagar** [comment]
Great contribution!

**Comment by dagar** [comment]
> This is really cool! Although, I do think it makes mores sense that the flightgear_bridge is a repository by itself and that it gets included as a submodule here, similarly to sitl_gazebo and jMavSim.

@TSC21 I think it's really a question of it being worth maintaining as a standalone project or not.
It very much is a standalone project, but realistically we might be better off maintaining it in tree unless someone wants to set up testing, CI, etc in a new repo.

**Comment by TSC21** [comment]
> > This is really cool! Although, I do think it makes mores sense that the flightgear_bridge is a repository by itself and that it gets included as a submodule here, similarly to sitl_gazebo and jMavSim.
> 
> @TSC21 I think it's really a question of it being worth maintaining as a standalone project or not.
> It very much is a standalone project, but realistically we might be better off maintaining it in tree unless someone wants to set up testing, CI, etc in a new repo.

I think the maintenance effort of something like this is independent if it is on a tree or not. It's more about the commitment to actually maintain it. And that needs to be the same being it on the Firmware tree or not. I just see thia being much more organized by being included in its own repo with its own structure and then having it included here as a submodule. I would expect thr CI pipeline for it tk be relatively easy to setup.

File: Tools/flightgear_bridge/src/fg_communicator.cpp [hunk 1]
```diff
@@ -0,0 +1,150 @@
+#include "fg_communicator.h"
+
+
+FGCommunicator::FGCommunicator(VehicleState * v)
```
Comment (dagar) [inline Tools/flightgear_bridge/src/fg_communicator.cpp]: You could run the c++ files through the code formatter.

``` Console
$ ./Tools/astyle/fix_code_style.sh Tools/flightgear_bridge/src/fg_communicator.cpp
```

File: Tools/flightgear_bridge/README [hunk 2]
```diff
@@ -0,0 +1,12 @@
+=== FlightGear Bridge ===
+
+Jak spusti vyvojovou verzi:
+
+1) na systému si zprovozněte FlightGear - klidně distribuční (tak aby v PATH bylo fgfs)
```
Comment (dagar) [inline Tools/flightgear_bridge/README]: Would you mind translating?

File: ROMFS/px4fmu_common/init.d-posix/17001_TF-G1 [hunk 3]
```diff
@@ -0,0 +1,57 @@
+#!/bin/sh
+#
+# @name Plane SITL
```
Comment (dagar) [inline ROMFS/px4fmu_common/init.d-posix/17001_TF-G1]: More specific name?

Comment (roman-dvorak) [reply ROMFS/px4fmu_common/init.d-posix/17001_TF-G1]: Renamed in d6ae722 
Added description d6ae7228f560beee834b3f7504bd45c3b28dc8b3

**Comment by Antiheavy** [comment]
I love flight gear!

File: Tools/flightgear_bridge/src/fg_communicator.cpp (hunk 1 above)
Comment (roman-dvorak) [reply Tools/flightgear_bridge/src/fg_communicator.cpp]: done in 09a69af, Thanks for the fix_code_style script link.

**Comment by slimonslimon** [comment]
At first thank you for extremly fast reaction.

How does it work:
 - It is stand alone app, which connect To FG (over UDP generic protocol) and PX4 ( over TCP, Mavlink)

There are few limitation, problems, and possible improvements:
 - as @kaklik mentioned, the gereric protocol of FG is somehow limitated by FPS of FG. So you get sensor data in low frequency, depending on your machine. (bridge can upsample it in terrible way). Question is: What is lower frequency of sensors data? I did not find that..
 - What do you thing about implementation of auto download of models for FlighGear?
 - There is strange problem with LockStep: FG build of PX4 need to be builded without LockStep, but I am not able to set up cmake scripts to build PX4 binary for FG without Lockstep and Gazebo Px4 Binary with LockStep. Can you help me? (https://github.com/PX4/Firmware/pull/14539/files#diff-a0bb2f464f832f23df5a4d5b1737cef2) -- I think this demage testes

File: Tools/flightgear_bridge/README (hunk 2 above)
Comment (slimonslimon) [reply Tools/flightgear_bridge/README]: Sorry...

File: ROMFS/px4fmu_common/mixers-sitl/autogyro_sitl.main.mix [hunk 4]
```diff
@@ -0,0 +1,37 @@
+Mixer for SITL plane
```
Comment (Jaeyoung-Lim) [inline ROMFS/px4fmu_common/mixers-sitl/autogyro_sitl.main.mix]: I think this is not a plane, or if it is identical with the plane can we use that instead?

File: Tools/sitl_run.sh [hunk 5]
```diff
@@ -121,6 +121,14 @@ elif [ "$program" == "gazebo" ] && [ ! -n "$no_sim" ]; then
 		echo "You need to have gazebo simulator installed!"
 		exit 1
 	fi
+elif [ "$program" == "flightgear" ] && [ ! -n "$no_sim" ]; then
+	echo "FG setup"
+    cd "${src_path}/Tools/flightgear_bridge/"
+    "${src_path}/Tools/flightgear_bridge/FG_run.py" ${model}.json
```
Comment (Jaeyoung-Lim) [inline Tools/sitl_run.sh]: What is the behavior if the ${model}.json doesn't exist? - or the model is not supported? It seems like only a plane and TF-G1 is supported. How can we handle people running e.g. `iris` more gracefully rather than doing something random?

File: platforms/posix/cmake/sitl_target.cmake [hunk 6]
```diff
@@ -57,12 +57,15 @@ ExternalProject_Add(mavsdk_tests
...
+add_subdirectory(${PX4_SOURCE_DIR}/Tools/flightgear_bridge ${PX4_BINARY_DIR}/build_flightgear_bridge EXCLUDE_FROM_ALL)
+
 # create targets for each viewer/model/debugger combination
-set(viewers none jmavsim gazebo)
+set(viewers none jmavsim gazebo flightgear)
 set(debuggers none ide gdb lldb ddd valgrind callgrind)
 set(models none shell
 	if750a iris iris_opt_flow iris_opt_flow_mockup iris_vision iris_rplidar iris_irlock iris_obs_avoid iris_rtps solo typhoon_h480
 	plane plane_cam plane_catapult
+	TF-G1
```
[the remark this answers is not in the document]
Comment (roman-dvorak) [reply platforms/posix/cmake/sitl_target.cmake]: Is it at least possible to leave it at ```tf_g1```? Full name of that UAV is [ThunderFly TF-G1](https://github.com/ThunderFly-aerospace/TF-G1).

File: ROMFS/px4fmu_common/mixers-sitl/autogyro_sitl.main.mix (hunk 4 above)
Comment (roman-dvorak) [reply ROMFS/px4fmu_common/mixers-sitl/autogyro_sitl.main.mix]: Yes, it is not same as plane mixer, fixed 9122b40a875762f5903cb142b73391ddb48791ac

File: Tools/sitl_run.sh (hunk 5 above)
Comment (slimonslimon) [reply Tools/sitl_run.sh]: well, I think that FG_run.py produce some error message (which will lost in output), and bridge will no start due to lack of parameters. PX4 starts but will be waiting for simulator forever.

But it seems that scripts stops after first error (I dont know why):
rover_error

We can check if backgrounded job FG_run.py is running or not after sleep command. Or have you better idea?
```

Comment (Jaeyoung-Lim) [reply Tools/sitl_run.sh]: @slimonslimon It would make sense to run the script for only a limited set of models since currently the list of supported models are quite small.

Comment (slimonslimon) [reply Tools/sitl_run.sh]: But the list of supported models is set by available .json files. Why should I have list of supported models somewhere and than configuration file for every model?

Or am I missing some configuration of available build targets?

Comment (slimonslimon) [reply Tools/sitl_run.sh]: when I run gazebo with unsuported model, it produce an error message about missing configuration file and open empty gazebo world, without model....

Comment (slimonslimon) [reply Tools/sitl_run.sh]: may we can improve CMake skript which generats available target.

Comment (Jaeyoung-Lim) [reply Tools/sitl_run.sh]: @slimonslimon Yes that is correct, but the unsupported models are not a make target. All models that are added as a target are supported in gazebo. The problem here is that some predefined make targets will just fail looking like it is just broken.

Comment (slimonslimon) [reply Tools/sitl_run.sh]: so it mean we should have list of supported models for every supported simulator (supported by make px4_sitl command) 

I will look at this later, because there is other problem because we need build px4 withou LOCKSTEP, and I dont know how to do it in CMake for group of targets (FG targets)

**Comment by slimonslimon** [comment]
Okey. We will move the bridge there.

But still remains the problem with cmake scripts and simulator LockStep and lists of supported vehicles per simulator. Any Idea?

**Comment by TSC21** [comment]
> Okey. We will move the bridge there.
> 
> But still remains the problem with cmake scripts and simulator LockStep and lists of supported vehicles per simulator. Any Idea?

We can solve that step by step. As far as I can tell this is WIP.

**Comment by slimonslimon** [comment]
Okey, I will prepare this step tomorow.

**Comment by hamishwillee** [comment]
This is cool. When this is done, we'll need some docs for environment setup, what is supported, how to use.

In the first instance, probably just a page like - http://dev.px4.io/master/en/simulation/gazebo.html but with setup information for different OS.

For reference, these are the main touchpoints for different platform setup docs/simulation:
- http://dev.px4.io/master/en/setup/dev_env.html#supported-targets
- http://dev.px4.io/master/en/setup/dev_env_linux_ubuntu.html#sim_nuttx
- http://dev.px4.io/master/en/simulation/
- http://dev.px4.io/master/en/simulation/gazebo.html
- http://dev.px4.io/master/en/simulation/hitl.html

**Comment by kaklik** [comment]
> @kaklik are you open that we create a repo under the PX4 org called PX4-FlightGear-Bridge, where you can push the bridge side and then include it here as a submodule? It's being seen as the correct architectural approach.

Ok, my opinion is that the best and easier way how to achieve this is the following:  We create a new repo named PX4-FlightGear-Bridge under our organization. You can immediately fork that repository in the PX4 organization.  Then we make pull-request or you itself should make a submodule link to your Firmware repository.

**Comment by TSC21** [comment]
> > @kaklik are you open that we create a repo under the PX4 org called PX4-FlightGear-Bridge, where you can push the bridge side and then include it here as a submodule? It's being seen as the correct architectural approach.
> 
> Ok, my opinion is that the best and easier way how to achieve this is the following: We create a new repo named PX4-FlightGear-Bridge under our organization. You can immediately fork that repository in the PX4 organization. Then we make pull-request or you itself should make a submodule link to your Firmware repository.

I don't see where that is easier: that will basically result in two more repositories to maintain, one by you and you only, and another (the PX4 fork) to be maintained by you and us. I think it is preferable that this is under am org where the maintenance responsibilities can be shared.

Regarding the submodule - that needs to be something part of this PR, so I ask you to do that yourselves. I will create the repo tomorrow and give you both write access to it (if I am able to, otherwise it will have to be someone with admin previlieges in the PX4 org). Then you can push the bridge there, if you agree.

**Comment by kaklik** [comment]
> > > @kaklik are you open that we create a repo under the PX4 org called PX4-FlightGear-Bridge, where you can push the bridge side and then include it here as a submodule? It's being seen as the correct architectural approach.
> > 
> > 
> > Ok, my opinion is that the best and easier way how to achieve this is the following: We create a new repo named PX4-FlightGear-Bridge under our organization. You can immediately fork that repository in the PX4 organization. Then we make pull-request or you itself should make a submodule link to your Firmware repository.
> 
> I don't see where that is easier: that will basically result in two more repositories to maintain, one by you and you only, and another (the PX4 fork) to be maintained by you and us. I think it is preferable that this is under an org where the maintenance responsibilities can be shared.
> 
> Regarding the submodule - that needs to be something part of this PR, so I ask you to do that yourselves. I will create the repo tomorrow and give you both write access to it (if I am able to, otherwise it will have to be someone with admin privileges in the PX4 org). Then you can push the bridge there, if you agree.

I see the pro and cons of both ways.  For us is mandatory to have multiple users access to the repository because few relatively random developers could experiment on branches in our repo.  Therefore we propabably could not avoid having a separate copy of that repository under our organization where we have privileges under our control.  The main problem which could emerge additionally from the fork of our repository to PX4 organization is the unclear location of the upstream repo for third party contributors. But it should be resolved by submodule path pointing to your fork of the repository under the PX4 organization. I saw you have some repositories included as a submodule in that way. The jMAVSim is one example.

**Comment by TSC21** [comment]
@kaklik I just take a look at your repo and seems that we might have a licensing issue, since you use GPL license and the PX4 ecosystem overall uses BSD-3clause or compatiblw.

File: platforms/posix/cmake/sitl_target.cmake (hunk 6 above)
[the remark this answers is not in the document]
Comment (kaklik) [reply platforms/posix/cmake/sitl_target.cmake:68]: What about `tf-g1` ? I suggest it because in other models in list underscore is used to denote the model variant. But the full name of our model is TF-G1 including the hyphen. (There are much more models in development including TF-G2, TF-G3..  etc.) The letter case does not matter for that.

**Comment by slimonslimon** [comment]
In last few commits I tried to solve problem with lockstep disabling for flightgear-bridge. I created a new label for px4_sitl board, which si copy of default, but disables lockstep of simulation. 
Then I improved sitl_targets.cmake, and allow gazebo and jMAVSim targets only if lockstep is enabled, and flightgear targets only if it is disabled.

I think, that gazebo and jMAVSim can work correctly without any other modification only if lockstep is enabled. Is it correct?

(and I fixed build of external project, after migration into module)

**Comment by TSC21** [comment]
@slimonslimon @kaklik are you planning to update your repo's license or not? I think you are ignoring something that truly signifies a Go or No-Go regarding this PR.

**Comment by slimonslimon** [comment]
@TSC21 I'm trying ti solve building and running problems these days. I'm a programmer not a loyer. I dont know much about licencnces.
(problably by using GitHub, you are agreeing with Microsoft EULA, and all codes shared here is property of Microsoft --- (joke) )

I think @kaklik will change licence as soon as he have time.

**Comment by kaklik** [comment]
> @kaklik I just took a look at your repo and seems that we might have a licensing issue, since you use GPL license and the PX4 ecosystem overall uses BSD-3clause or compatible. I am not sure we can include it as a submodule that way.

We just changed the repo license to BSD-3Clause.

**Comment by TSC21** [comment]
> @TSC21 I'm trying ti solve building and running problems these days. I'm a programmer not a lawyer. I dont know much about licencnces.
> (problably by using GitHub, you are agreeing with Microsoft EULA, and all codes shared here is property of Microsoft --- (joke) )
> 
> I think @kaklik will change licence as soon as he have time.

Now arguing, but a programmer should be aware of SW licensing as well or it might run into problems eventually ;)

**Comment by TSC21** [comment]
> > @kaklik I just took a look at your repo and seems that we might have a licensing issue, since you use GPL license and the PX4 ecosystem overall uses BSD-3clause or compatible. I am not sure we can include it as a submodule that way.
> 
> We just changed the repo license to BSD-3Clause.

Thanks!

File: platforms/posix/cmake/sitl_target.cmake [hunk 7]
```diff
@@ -57,94 +58,161 @@ ExternalProject_Add(mavsdk_tests
...
+    set(debuggers none ide gdb lldb ddd valgrind callgrind)
+    set(models none shell
+	    plane
+        )
+    set(worlds none empty)
+    set(all_posix_vmd_make_targets)
+    foreach(viewer ${viewers})
+	    foreach(debugger ${debuggers})
+		    foreach(model ${models})
+			    foreach(world ${worlds})
```
Comment (Jaeyoung-Lim) [inline platforms/posix/cmake/sitl_target.cmake:176]: You are not using any notion of worlds, no need to iterate on worlds

File: platforms/posix/cmake/sitl_target.cmake [hunk 8]
```diff
@@ -57,94 +58,161 @@ ExternalProject_Add(mavsdk_tests
...
+    endforeach()
+else()
+    set(viewers none flightgear)
+    set(debuggers none ide gdb lldb ddd valgrind callgrind)
+    set(models none shell
+	    plane
+        )
+    set(worlds none empty)
+    set(all_posix_vmd_make_targets)
+    foreach(viewer ${viewers})
```
Comment (Jaeyoung-Lim) [inline platforms/posix/cmake/sitl_target.cmake:173]: There is only one viewer (flight gear), no need to iterate on viewers

File: platforms/posix/cmake/sitl_target.cmake [hunk 9]
```diff
@@ -57,94 +58,161 @@ ExternalProject_Add(mavsdk_tests
...
+	    foreach(debugger ${debuggers})
```
Comment (Jaeyoung-Lim) [inline platforms/posix/cmake/sitl_target.cmake:174]: Have you tested with each of the debugger options? If not I suggest we just remove it until it is tested

File: platforms/posix/cmake/sitl_target.cmake [hunk 10]
```diff
@@ -57,94 +58,161 @@ ExternalProject_Add(mavsdk_tests
...
+						    list(APPEND all_posix_vmd_make_targets ${_targ_name})
+						    add_dependencies(${_targ_name} px4 sitl_gazebo)
+					    endif()
+				    endif()
+			    endforeach()
+		    endforeach()
+	    endforeach()
+    endforeach()
+else()
+    set(viewers none flightgear)
```
Comment (Jaeyoung-Lim) [inline platforms/posix/cmake/sitl_target.cmake:166]: Since there is only flight gear, you don't need to reset viewers as a variable

File: platforms/posix/cmake/sitl_target.cmake [hunk 11]
```diff
@@ -57,94 +58,161 @@ ExternalProject_Add(mavsdk_tests
...
+    set(debuggers none ide gdb lldb ddd valgrind callgrind)
+    set(models none shell
+	    plane
+        )
+    set(worlds none empty)
```
Comment (Jaeyoung-Lim) [inline platforms/posix/cmake/sitl_target.cmake:171]: Worlds are not being used. No need to iterate on worlds

File: platforms/posix/cmake/sitl_target.cmake [hunk 12]
```diff
@@ -57,94 +58,161 @@ ExternalProject_Add(mavsdk_tests
...
+	DEPENDS
+		git_flightgear_bridge
+	USES_TERMINAL_CONFIGURE true
+	USES_TERMINAL_BUILD true
+	EXCLUDE_FROM_ALL true
+	BUILD_ALWAYS 1
+)
+
+
+if( ENABLE_LOCKSTEP_SCHEDULER STREQUAL "yes")
```
Comment (Jaeyoung-Lim) [inline platforms/posix/cmake/sitl_target.cmake:76]: Can you elaborate on this condition? why do we need it? we still need to be able to run jmavsim and gazebo without lockstep

File: boards/px4/sitl/default.cmake [hunk 13]
```diff
@@ -102,5 +102,5 @@ if(REPLAY_FILE)
 	message(STATUS "Building without lockstep for replay")
 	set(ENABLE_LOCKSTEP_SCHEDULER no)
 else()
-	set(ENABLE_LOCKSTEP_SCHEDULER yes)
+	set(ENABLE_LOCKSTEP_SCHEDULER yes)
```
Comment (Jaeyoung-Lim) [inline boards/px4/sitl/default.cmake:105]: Please remove the white space

File: ROMFS/px4fmu_common/init.d-posix/17001_TF-G1 [hunk 14]
```diff
@@ -0,0 +1,63 @@
+#!/bin/sh
+#
+# @name ThunderFly TF-G1
+# ThunderFly TF-G1 autogyro airframe. Only for FlightGear simulator
+#
+# @type Autogyro
+# @class Autogyro
+#
+# @url https://github.com/ThunderFly-aerospace/TF-G1/
```
Comment (Jaeyoung-Lim) [inline ROMFS/px4fmu_common/init.d-posix/17001_TF-G1:9]: I understand the model name is in captial letters, but adding this as a make target breaks the convention. No other model has a capital letter in the target name. For example, if750a is supposed to be IF750A but it is written in small case letters. Would it be possible to change this?

File: platforms/posix/cmake/sitl_target.cmake (hunk 12 above)
Comment (slimonslimon) [reply platforms/posix/cmake/sitl_target.cmake:76]: Thank you - I will clear it.

Comment (slimonslimon) [reply platforms/posix/cmake/sitl_target.cmake:76]: Done. sitl_target.cmake  simlified and gazebo and etc added back for build without lockstep.

File: boards/px4/sitl/default.cmake (hunk 13 above)
Comment (slimonslimon) [reply boards/px4/sitl/default.cmake:105]: Sorry, done.

File: platforms/posix/cmake/sitl_target.cmake [hunk 15]
```diff
@@ -57,95 +58,138 @@ ExternalProject_Add(mavsdk_tests
...
-							logs_symlink
-						)
-						list(APPEND all_posix_vmd_make_targets ${_targ_name})
-						add_dependencies(${_targ_name} px4 sitl_gazebo)
-					endif()
-				endif()
-			endforeach()
-		endforeach()
-	endforeach()
+    foreach(debugger ${debuggers})
```
Comment (Jaeyoung-Lim) [inline platforms/posix/cmake/sitl_target.cmake:87]: Could you please remove all the whitespace changes? or is this intended?

File: platforms/posix/cmake/sitl_target.cmake (hunk 6 above)
[the remark this answers is not in the document]
Comment (slimonslimon) [reply platforms/posix/cmake/sitl_target.cmake]: Renamed to tf-g1

File: platforms/posix/cmake/sitl_target.cmake (hunk 12 above)
Comment (Jaeyoung-Lim) [reply platforms/posix/cmake/sitl_target.cmake]: Thanks! Looks much better now

File: Tools/sitl_run.sh (hunk 5 above)
Comment (slimonslimon) [reply Tools/sitl_run.sh:127]: both "Lockstep" and "list of models! are solved

File: platforms/posix/cmake/sitl_target.cmake (hunk 15 above)
Comment (slimonslimon) [reply platforms/posix/cmake/sitl_target.cmake]: okey, I tried this.

Comment (slimonslimon) [reply platforms/posix/cmake/sitl_target.cmake]: Done.

File: ROMFS/px4fmu_common/init.d-posix/17001_TF-G1 (hunk 14 above)
Comment (slimonslimon) [reply ROMFS/px4fmu_common/init.d-posix/17001_TF-G1]: renamed to tf-g1 in last commits

File: Tools/sitl_run.sh [hunk 16]
```diff
@@ -121,6 +121,14 @@ elif [ "$program" == "gazebo" ] && [ ! -n "$no_sim" ]; then
 		echo "You need to have gazebo simulator installed!"
 		exit 1
 	fi
+elif [ "$program" == "flightgear" ] && [ ! -n "$no_sim" ]; then
```
Comment (bkueng) [inline Tools/sitl_run.sh:124]:
```suggestion
elif [ "$program" == "flightgear" ] && [ -z "$no_sim" ]; then
```

File: Tools/sitl_run.sh [hunk 17]
```diff
@@ -121,6 +121,14 @@ elif [ "$program" == "gazebo" ] && [ ! -n "$no_sim" ]; then
...
+	echo "FG setup"
+    cd "${src_path}/Tools/flightgear_bridge/"
+    "${src_path}/Tools/flightgear_bridge/FG_run.py" ${model}.json    
+    sleep 20
```
Comment (bkueng) [inline Tools/sitl_run.sh:128]: Do you really need to wait that long?

Comment (slimonslimon) [reply Tools/sitl_run.sh:128]: Now, we have not any implementation of lock/check if FG is inicialized. And there is a problem, that, FG generic protocol sends bad data befere FG is fully inicialized.. 
So create this long sleep workround..

I will search for some check if FG is ready...

File: ROMFS/px4fmu_common/init.d-posix/1061_tf-r1 [hunk 18]
```diff
@@ -0,0 +1,43 @@
+#!/bin/sh
+#
+# @name ThunderFly TF-R1 UAV Rover
+# @type Rover
+# @class Rover
```
Comment (Jaeyoung-Lim) [inline ROMFS/px4fmu_common/init.d-posix/1061_tf-r1:5]: Is this model also related to flight gear support?

Comment (kaklik) [reply ROMFS/px4fmu_common/init.d-posix/1061_tf-r1:5]: This model is compatible with PX4 rover support.

File: Tools/sitl_run.sh (hunk 17 above)
Comment (slimonslimon) [reply Tools/sitl_run.sh]: Well, in last commit We solved this issue. There is no more sleep now and FG-bridge start send messages when simulation time is large than zero.

Comment (bkueng) [reply Tools/sitl_run.sh]: Cool

File: Tools/setup/ubuntu.sh [hunk 19]
```diff
@@ -203,6 +203,10 @@ if [[ $INSTALL_SIM == "true" ]]; then
 		protobuf-compiler \
 		;
 
+	# FlightGear
+	sudo add-apt-repository -y -u ppa:saiarcot895/flightgear
+	sudo DEBIAN_FRONTEND=noninteractive apt-get -y --quiet install flightgear
```
Comment (Jaeyoung-Lim) [inline Tools/setup/ubuntu.sh:208]: In my experience, Flight gear takes quite some time to install, and I don't think this should be the default. Could we separate this out?

Comment (kaklik) [reply Tools/setup/ubuntu.sh:208]: ok, what is the best way to separate flightgear installation?

**Comment by Jaeyoung-Lim** [comment]
@slimonslimon Is this ready to get in? Since there has been a lot of updates since the last review, not sure if this is ready to get in

**Comment by kaklik** [comment]
> @slimonslimon Is this ready to get in? Since there has been a lot of updates since the last review, the reviews are getting dismissed. Once this looks stable on your side, I will test it and maybe we can merge it?

The changes which we continually adding are the bug fixes of issues, which we found by ourselves.  But the pull-request is all time working and stable from our side.  I think it should be merged because we probably want to add new features that should be part of another pull-requests.

The actual important step which I feel as missing is documentation of the current code. As @hamishwillee [noted in his comment](https://github.com/PX4/Firmware/pull/14539#issuecomment-606883322).

**Comment by Jaeyoung-Lim** [comment]
@kaklik Awesome, thanks. Could you maybe squash the commits so that the commit history is more sensible? (There are 56 commits and most of them contain redundant updates or submodule updates)

Also, what are the plans for documentation? Would you be able to take care of the documentation? I would appreciate one, since my experience on bringing up flightgear was not smooth last time.

**Comment by Jaeyoung-Lim** [comment]
@kaklik @slimonslimon It is really cool that you are improving the flight gear integration.

However, could we freeze this branch and do squash /cleanups so that we can merge this branch?

**Comment by slimonslimon** [comment]
@Jaeyoung-Lim, yesterday we spend lot of time, by debuging the FG connection, but the result is, that problems are probably not in bridge but in FG. ( Flight gear is not so good :-( ).

 I will do the squash of commits at this weekend.

**Comment by Jaeyoung-Lim** [comment]
@kaklik @slimonslimon I have tried the rascal vehicle with flightgear 2019.1.1 on Ubuntu bionic. 

However, it results in a flyaway in misison mode. 
[Demo](https://youtu.be/EEpRw_IFSNU)

Is this the expected outcome?

**Comment by kaklik** [comment]
>  It results in a flyaway in mission mode. (See video below) Is this the expected outcome? what would be the correct way to test this?



We noticed this also, but it seems to be not related to FlightGear connection. It is probably some bug in the PX4 mission solving algorithm.  The expected behavior is the following: 

[FlightGear SITL Plane test](https://www.youtube.com/watch?v=iqdcN5Gj4wI&feature=youtu.be)

>  
>     * On startup of the simulation, the vehicle is moving. Is this intended?
> 

Yes we now that, you have probably used the Rascal model with a combustion engine.  The thrust generated by the propeller in idle RPM is large enough to make airplane moving.  It could be solved in case PX4 is able to handle gear brakes.  But this is not available as I know.  The current solution is use of model [with electric engine](https://github.com/ThunderFly-aerospace/FlightGear-Rascal).

**Comment by Jaeyoung-Lim** [comment]
@kaklik How can I use the electric engine model?

**Comment by roman-dvorak** [comment]
> How can I use the electric engine model?
The gas and the electric motor has a different make target. 

Combustion engine:
`make px4_sitl_nolockstep flightgear_rascal`

Electric motor:
`make px4_sitl_nolockstep flightgear_rascal-electric`

The electric motor support is in master since yesterday. So you need to install the latest daily-build of FlightGear.

**Comment by Jaeyoung-Lim** [comment]
@roman-dvorak Okay, where can I get the latest daily-build of flight gear? Please forgive me if these are stupid questions, but as I am not familiar with flightgear, it feels like there a lot of moving parts. 

Could you document the exact versioning / system that this has been tested in the readme?

**Comment by kaklik** [comment]
> @roman-dvorak Okay, where can I get the latest daily-build of flight gear? Please forgive me if these are stupid questions, but as I am not familiar with flightgear, it feels like there are a lot of moving parts.
> 
> Could you document the exact versioning / system that this has been tested in the readme?

The location where you get the daily build of FG is pointed from the [Rascal model readme](https://github.com/ThunderFly-aerospace/FlightGear-Rascal)

**Comment by slimonslimon** [comment]
@Jaeyoung-Lim, @kaklik is preparing more detailed documentation for documentation web pages, because it can be maintainted in better, than readme in repo (I think). The readme will be updatet soon by @kaklik .

There are lots of problems with FlightGear (versions, data-versions, lots of parameters, undocumended feautres, documentation by some posts on wiki, or forum, single thread implemntation from previous century,  etc.). We probably will not be able to write down good and complente documentation for flightgear. But We maybe will try wrote some Introduction to flightgear along with documentation of using bridge with PX4.

**Comment by Jaeyoung-Lim** [comment]
@slimonslimon Thanks, I totally agree. We don't need to document flight gear. I think for a start, a "bullet proof" step by step process to get it up and running will be already really good. I seem to keep running in to problems while depending on the readme of the px4-flightgear-bridge.

Anyway, I would like to get this in @dagar @TSC21 would it be possible to merge this to move forward?

File: Tools/setup/ubuntu.sh (hunk 19 above)
Comment (bkueng) [reply Tools/setup/ubuntu.sh:205]: @Jaeyoung-Lim @kaklik this is still unanswered, and I agree, it should be optional / a separate script.

File: Tools/sitl_run.sh [hunk 20]
```diff
@@ -121,6 +121,13 @@ elif [ "$program" == "gazebo" ] && [ ! -n "$no_sim" ]; then
 		echo "You need to have gazebo simulator installed!"
 		exit 1
 	fi
+elif [ "$program" == "flightgear" ] && [ -z "$no_sim" ]; then
+	echo "FG setup"
+    cd "${src_path}/Tools/flightgear_bridge/"
```
Comment (bkueng) [inline Tools/sitl_run.sh:126]: Indentation: use tabs (below too)

File: platforms/posix/cmake/sitl_target.cmake [hunk 21]
```diff
@@ -146,6 +161,38 @@ foreach(viewer ${viewers})
 	endforeach()
 endforeach()
 
+#add flighgear targets
+if( ENABLE_LOCKSTEP_SCHEDULER STREQUAL "no")
+    set(models
```
Comment (bkueng) [inline platforms/posix/cmake/sitl_target.cmake:166]: Indentation: tabs

**Comment by slimonslimon** [comment]
@bkueng I think, it is not posilbe to setup lockstep simulation with FlightGear by this way (communication over generic protocol), because I dont know how to stop the simulation of FG. So we decided to create _nolockstep target and connect FG targets to it.

File: Tools/setup/ubuntu.sh (hunk 19 above)
Comment (kaklik) [reply Tools/setup/ubuntu.sh]: I have removed it from the script. Installation will be explained in the Devguide documentation.

File: Tools/sitl_run.sh (hunk 20 above)
Comment (slimonslimon) [reply Tools/sitl_run.sh]: okey.

File: platforms/posix/cmake/sitl_target.cmake (hunk 21 above)
Comment (slimonslimon) [reply platforms/posix/cmake/sitl_target.cmake]: done.

**Comment by slimonslimon** [comment]
I hope I fixed every 4space->tab. 

Then we added model into bridge as submodules. It should help to do "fisrt run" as simple as posible. Some notes are in updated README of Bridge. More Detailed documantation is under developement by @kaklik

**Comment by kaklik** [comment]
Documentation of FlightGear support is now available in [PX4Devguide](https://dev.px4.io/master/en/simulation/flightgear.html)
